src/probe/probe.c

```diff
@@ -15,42 +24,551 @@
+		return err;
+	}
+
+	dma_channel_put(dma->dc.chan);
+
+	rfree((void *)dma->dmapb.addr);
+	dma->dmapb.addr = 0;
+
+	dma->stream_tag = PROBE_DMA_INVALID;
+
+	return 0;
+}
+
 int probe_init(struct probe_dma *probe_dma)
 {
+	uint32_t i;
+	int err;
+
+	tracev_probe("probe_init()");
+
+	if (_probe) {
+		trace_probe_error("probe_init() error: Probes already initialized.");
+		return -EINVAL;
+	}
+
+	/* alloc probes main struct */
+	_probe = rzalloc(SOF_MEM_ZONE_SYS_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+			 sizeof(*_probe));
+
+	if (!_probe) {
+		trace_probe_error("probe_init() error: Alloc failed.");
+		return -ENOMEM;
+	}
+
+	/* setup extraction dma if requested */
+	if (probe_dma) {
+		tracev_probe("\tstream_tag = %u, dma_buffer_size = %u",
+			     probe_dma->stream_tag, probe_dma->dma_buffer_size);
+
+		_probe->ext_dma.stream_tag = probe_dma->stream_tag;
+		_probe->ext_dma.dma_buffer_size = probe_dma->dma_buffer_size;
+
+		err = probe_dma_init(&_probe->ext_dma);
+		if (err < 0) {
+			trace_probe_error("probe_init() error: probe_dma_init() failed");
+			_probe->ext_dma.stream_tag = PROBE_DMA_INVALID;
+			return err;
+		}
+	} else {
+		tracev_probe("\tno extraction DMA setup");
+
+		_probe->ext_dma.stream_tag = PROBE_DMA_INVALID;
+	}
+
+	/* initialize injection DMAs as invalid */
+	for (i = 0; i < CONFIG_PROBE_DMA_MAX; i++)
+		_probe->inject_dma[i].stream_tag = PROBE_DMA_INVALID;
+
+	/* initialize probe points as invalid */
+	for (i = 0; i < CONFIG_PROBE_POINTS_MAX; i++)
+		_probe->probe_points[i].stream_tag = PROBE_POINT_INVALID;
+
 	return 0;
 }
 
 int probe_deinit(void)
 {
+	uint32_t i;
+	int err;
+
+	tracev_probe("probe_deinit()");
+
+	if (!_probe) {
+		trace_probe_error("probe_deinit() error: Not initialized.");
+
+		return -EINVAL;
+	}
+
+	/* check for attached injection probe DMAs */
+	for (i = 0; i < CONFIG_PROBE_DMA_MAX; i++) {
+		if (_probe->inject_dma[i].stream_tag != PROBE_DMA_INVALID) {
+			trace_probe_error("probe_deinit() error: Cannot deinitialize with injection DMAs attached.");
+			return -EINVAL;
+		}
+	}
+
+	/* check for connected probe points */
+	for (i = 0; i < CONFIG_PROBE_POINTS_MAX; i++) {
+		if (_probe->probe_points[i].stream_tag != PROBE_POINT_INVALID) {
+			trace_probe_error("probe_deinit() error: Cannot deinitialize with probe points connected.");
+			return -EINVAL;
+		}
+	}
+
+	if (_probe->ext_dma.stream_tag != PROBE_DMA_INVALID) {
+		tracev_probe("probe_deinit() Freeing extraction DMA.");
+		err = probe_dma_deinit(&_probe->ext_dma);
+		if (err < 0)
+			return err;
+	}
+
+	rfree(_probe);
+	_probe = NULL;
+
 	return 0;
 }
 
 int probe_dma_add(uint32_t count, struct probe_dma *probe_dma)
 {
+	uint32_t i;
+	uint32_t j;
+	uint32_t stream_tag;
+	uint32_t first_free;
+	int err;
+
+	tracev_probe("probe_dma_add() count = %u", count);
+
+	if (!_probe) {
+		trace_probe_error("probe_dma_add() error: Not initialized.");
+
+		return -EINVAL;
+	}
+
+	for (i = 0; i < count; i++) {
```

Comment:
Why are we looping here, inline comment.

Reply:
added